File: qt/aqt/retrievability.py
# Copyright: Ankitects Pty Ltd and contributors
# License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

"""The word "retrievability" explains itself on hover wherever a Qt screen
shows it (spec ui.retrievability-advanced-only): underlined in a label, with
"probability of recall" as the tooltip. The screens that show it are
Advanced-only. The web pages do the same with RetrievabilityText.svelte."""

from __future__ import annotations

import html
import re

from aqt.qt import QLabel, Qt
from aqt.utils import tr


def explanation() -> str:
    """What the hover says."""
    return tr.card_stats_retrievability_explanation()


def _term_pattern() -> re.Pattern[str] | None:
    term = tr.card_stats_retrievability_term()
    return re.compile(re.escape(term), re.IGNORECASE) if term else None


def mentions_retrievability(text: str) -> bool:
    pattern = _term_pattern()
    return bool(pattern and pattern.search(text))


def underlined_html(text: str) -> str:
    """`text` as rich text, every occurrence of the word underlined."""
    escaped = html.escape(text)
    pattern = _term_pattern()
    if pattern is None:
        return escaped
    return pattern.sub(lambda match: f"<u>{match.group(0)}</u>", escaped)


def explain_in_label(label: QLabel, text: str) -> None:
    """Shows `text` in `label`, the word underlined and explained on hover."""
    label.setTextFormat(Qt.TextFormat.RichText)
    label.setText(underlined_html(text))
    label.setToolTip(explanation() if mentions_retrievability(text) else "")

File: ts/lib/tslib/retrievability.ts
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

/**
 * The word "retrievability" explains itself on hover wherever a text shows
 * it (spec ui.retrievability-advanced-only): it is underlined, and the hover
 * says "probability of recall". The texts that show it are Advanced-only.
 */

import * as tr from "@generated/ftl";

/** A piece of a text: the explained word, or the text around it. */
export interface TextPart {
    text: string;
    term: boolean;
}

/** `text` cut around every occurrence of `term`, whatever its case. A text
 * without it is one plain part, and so is every text when `term` is empty. */
export function termParts(text: string, term: string): TextPart[] {
    const parts: TextPart[] = [];
    const needle = term.toLocaleLowerCase();
    const haystack = text.toLocaleLowerCase();
    let from = 0;
    while (needle) {
        const at = haystack.indexOf(needle, from);
        if (at < 0) {
            break;
        }
        if (at > from) {
            parts.push({ text: text.slice(from, at), term: false });
        }
        parts.push({ text: text.slice(at, at + needle.length), term: true });
        from = at + needle.length;
    }
    if (from < text.length) {
        parts.push({ text: text.slice(from), term: false });
    }
    return parts;
}

/** `text` cut around the word "retrievability" of the current language. */
export function retrievabilityParts(text: string): TextPart[] {
    return termParts(text, tr.cardStatsRetrievabilityTerm());
}

/** What the hover says. */
export function retrievabilityExplanation(): string {
    return tr.cardStatsRetrievabilityExplanation();
}
